### PR TITLE
BHBC-992: Log out / Log in

### PR DESCRIPTION
**Pull request by NickPhura**
# Overview

## Links to jira tickets

https://quartech.atlassian.net/browse/BHBC-992

## This PR contains the following changes

Added logout util function to log user out of siteminder and keycloak
Added log in page
Updated routing to allow user to reach log in page (when not authenticated).
- Previously, ALL routes required authentication
Minor misc updates to doc/naming around routing/auth
Removed AuthLayout, as it is/was not doing anything useful
Added user name and logout link to header

## This PR contains the following types of changes

- [x] New feature (change which adds functionality)
- [ ] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [x] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.

### CODE REVIEW

File: app/src/pages/200/RequestSubmitted.tsx [hunk 1]
```diff
@@ -1,15 +1,36 @@
...
+  }
+
+  if (keycloakWrapper?.systemRoles.length) {
+    // User already has a role
+    return <Redirect to={{ pathname: '/' }} />;
+  }
+
+  if (!keycloakWrapper.hasAccessRequest) {
+    // User has no pending access request
+    return <Redirect to={{ pathname: '/' }} />;
```
Comment (sdevalapurkar) [inline app/src/pages/200/RequestSubmitted.tsx:31]: question:
how come we redirect to the same place on both these cases?

Comment (sdevalapurkar) [reply app/src/pages/200/RequestSubmitted.tsx:31]: or is this a todo that needs to be updated later?

File: app/src/utils/Utils.ts [hunk 2]
```diff
@@ -79,3 +80,22 @@ export const getFormattedAmount = (amount: number): string => {
       .replace(/\B(?=(\d{3})+(?!\d))/g, ',')
   );
 };
+
+/**
+ * Log the user out, redirecting them to the logout page.
```
Comment (sdevalapurkar) [inline app/src/utils/Utils.ts:85]: question: should this say
```suggestion
 * Log the user out, redirecting them to the login page.
```

**Review by sdevalapurkar — COMMENTED** [review]
lgtm, few questions though - just wanted to clarify before approval :)

File: app/src/components/layout/Header.tsx [hunk 3]
```diff
@@ -77,35 +97,82 @@ const useStyles = makeStyles((theme: Theme) => ({
...
-            <Link to="/projects" color={'inherit'} id="menu_projects">
-              Projects
-            </Link>
-            {keycloakWrapper?.hasSystemRole([SYSTEM_ROLE.SYSTEM_ADMIN]) && (
-              <Link to="/admin/users" color={'inherit'} id="menu_admin_users">
-                Manage Users
-              </Link>
-            )}
+            <SecureLink to="/projects" label="Projects" id="menu_projects" />
+            <SecureLink to="/admin/users" label="Mange Users" id="menu_admin_users" />
```
Comment (anissa-agahchen) [inline app/src/components/layout/Header.tsx]: label="Manage Users

**Comment by anissa-agahchen** [comment]
hmm ... submitting an access request doesn't work ... can you merge dev into it? .. might fix it.

File: app/src/pages/200/RequestSubmitted.tsx (hunk 1 above)
Comment (NickPhura) [reply app/src/pages/200/RequestSubmitted.tsx:31]: The redirect is the same, I just kept the checks separate for readability.  There are similar if checks in a few places and I was just keeping them separate with comments to make sure its clear whats going on

**Comment by anissa-agahchen** [comment]
> hmm ... submitting an access request doesn't work ... can you merge dev into it? .. might fix it.

I suspect there is an issue with one of the useeffects ....submitting an access-request works fine, but sends you back to fill in another request.  If you refresh though, it correctly re-routes to the request-submitted page.

File: app/src/components/layout/Header.tsx (hunk 3 above)
Comment (NickPhura) [reply app/src/components/layout/Header.tsx]: Fixed

File: app/src/utils/Utils.ts (hunk 2 above)
Comment (NickPhura) [reply app/src/utils/Utils.ts:85]: Fixed

**Review by sdevalapurkar — DISMISSED** [review]
code lgtm, need to run locally

File: app/src/pages/login/LogInPage.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    // User has a role
+    return <Redirect to={{ pathname: '/projects' }} />;
+  }
+
+  return (
+    <Container>
+      <Box pt={6} textAlign="center">
+        <Icon path={mdiDoorClosedLock} size={2} className={classes.icon} />
+        <h1>Logged Out</h1>
```
Comment (NickPhura) [inline app/src/pages/login/LogInPage.tsx]:
```suggestion
        <h1>Welcome to BioHub</h1>
```

**Review by sdevalapurkar — DISMISSED** [review]

**Review by sdevalapurkar — APPROVED** [review]

**Pull request merged**